src/storage/src/render/persist_sink.rs

```diff
@@ -932,439 +935,574 @@ where
                 }
-            });
+            })
+            .collect::<Vec<_>>();
+
+        trace!(
+            "persist_sink {collection_id}/{shard_id}: \
+                    append_batches: in_flight_descriptions size: {:?}, \
+                    in_flight_batches size: {:?}, \
+                    done size: {:?}, \
+                    batch_frontier: {:?}, \
+                    batch_description_frontier: {:?}",
+            in_flight_descriptions.len(),
+            in_flight_batches.len(),
+            done_batches.len(),
+            batches_frontier,
+            batch_description_frontier
+        );
 
-            // Reverse, as we'll pop batches off the end of the queue.
-            done_batches.reverse();
+        let result = update_persist_with_batches(
+            collection_id,
+            worker_id,
+            shard_id,
+            &metrics,
+            &busy_signal,
+            read_only_rx,
+            bail_on_concurrent_modification,
+            write,
+            upper_cap_set,
+            &current_upper,
+            &source_statistics,
+            &mut done_batches,
+            &mut in_flight_batches,
+            &mut in_flight_descriptions,
+        )
+        .await;
+
+        if let Err(e) = result {
+            return Err(e);
+        }
+    }
+}
 
-            while let Some(done_batch_metadata) = done_batches.pop() {
-                in_flight_descriptions.remove(&done_batch_metadata);
+async fn update_persist_with_batches(
+    collection_id: GlobalId,
+    worker_id: usize,
+    shard_id: ShardId,
+    metrics: &SourcePersistSinkMetrics,
+    busy_signal: &Arc<Semaphore>,
+    read_only_rx: &mut watch::Receiver<bool>,
+    bail_on_concurrent_modification: bool,
+    write: &mut mz_persist_client::write::WriteHandle<
+        SourceData,
+        (),
+        mz_repr::Timestamp,
+        StorageDiff,
+    >,
+    upper_cap_set: &mut CapabilitySet<mz_repr::Timestamp>,
+    current_upper: &Rc<std::cell::RefCell<Antichain<mz_repr::Timestamp>>>,
+    source_statistics: &StorageStatistics<
+        SourceStatisticsRecord,
+        SourceStatisticsMetrics,
+        SourceStatisticsMetadata,
+    >,
+    done_batches: &mut Vec<(Antichain<mz_repr::Timestamp>, Antichain<mz_repr::Timestamp>)>,
+    in_flight_batches: &mut HashMap<
+        (Antichain<mz_repr::Timestamp>, Antichain<mz_repr::Timestamp>),
+        BatchSet,
+    >,
+    in_flight_descriptions: &mut std::collections::HashSet<(
+        Antichain<mz_repr::Timestamp>,
+        Antichain<mz_repr::Timestamp>,
+    )>,
+) -> Result<(), anyhow::Error> {
+    let (mut batch_lower, batch_upper) = match (done_batches.first(), done_batches.last()) {
+        (Some(first), Some(last)) => (first.0.clone(), last.1.clone()),
+        (None, None) => {
+            upper_cap_set.downgrade(current_upper.borrow().iter());
+            //
+            return Ok(());
+        }
+        // SAFETY(ptravers): if first exists so must last.
```

Comment:
`SAFETY` comments are reserved for documenting how unsafe code blocks are upholding whatever invariants are necessary to avoid UB. Since there is no `unsafe` here we should make this a normal comment.